Implement Sensor API on the controller

amcgee opened the pull request:
This API allows sensor modules to push values to the controller.  The values are then durably logged to Flash memory (a flash_queue ring buffer)  This also refactors memory_manager, momo_config, and flash_queue to be better and utilize the persistent flashblock implementation.

Some implementation notes:
- The first subsection of the flash_queue is now reserved for a flashblock storing the ringbuffer pointers
- Currently, writing values and the flashbock to memory is done synchronously in the MIB callback - this should be asynchronous and probably utilize a small RAM buffer array (in sensor_manager.c).
- I have an adaptive reporting implementation (which aggregates intelligently based on verbosity settings and the actual data to be reported) but I'll put that in a separate pull request to (hopefully) not confuse things as much. (Issue #27)
- Implementation of a sensor module registry (through module registrations) is still pending.
- I initially had a rudimentary "schedule value polling" implementation, but I decided it would be better to do a generic "schedule this callback every x interval" on the controller, and then have the module use the normal "push" API to record a value.
- I've added a python tool called `momosensor` which can:
  - `log` individual values
  - `readone` value, popping from the front of the values queue
  - `emulate` to write a random value every .1 seconds.

Sorry this diff is so large.

Closes #65

amcgee committed:
Initial sensor API implementation.

amcgee committed:
Basic test harness, some bug fixes.

amcgee committed:
Fix persistent-state bug, confirmed working with test harness.

amcgee committed:
Create separate sensor emulator tool to log values repeatedly.

amcgee committed:
Significantly reduce memory use by (correctly) using uint32 instead of uint64 for memory addresses.  Also save off start and end addresses, rather than computing them from subsection indices every time.  Fix a bug introduced when memory_manager changed.

amcgee committed:
Fix some bugs and make the momosensor proxy more robust.

amcgee committed:
Remove hard-coded 11 address from GSMTool, use cmdln for argument parsing.

amcgee committed:
Make the metadata type explicit.

amcgee committed:
Fix typo causing build failure.

amcgee committed:
Fix emulation logic to accept a sensor address and pass 0 for metadata by default.

amcgee committed:
We don't need the old sensor code (yay version control\!) so let's just get rid of it.

amcgee committed:
Remove a silly unnecessary line ending change from report_manager.h

amcgee committed:
Use 32-bit values instead of 64 - if a sensor really needs more that 2^32 options for a single reading value they can do a custom aggregation of two streams or something.  In general most things shouldn't need more than 2^32 precision (if even that)

amcgee committed:
Merge branch 'sensor-api' of https://github.com/amcgee/MoMo-Firmware into sensor-api

amcgee committed:
Fix momosensor tool to expect a 32-bit value.

amcgee commented:
Changed value size from 64-bit to 32-bit.

NOTE:  If there are no sensor events to read, the `momosensor readone` command will indicate a code 7 callback error because that's what the read_sensor_log endpoint throws.

amcgee committed:
Fix sensor log overrun bug.

timburke commented on momo_modules/shared/portable/core/common_types.h:
```diff
@@ -10,10 +10,18 @@ typedef unsigned short bool;
+
 typedef signed int      int16;
 typedef unsigned int    uint16;
+#define MAX_UINT16 65,535
```
You can't have commas in int literals.  It will be macro expanded and then C will see a comma operator.  It may still compile but will not work correctly.

amcgee replied:
Good catch yeah, I added these but they aren't actually used.

timburke commented on momo_modules/shared/portable/core/common_types.h:
```diff
@@ -10,10 +10,18 @@ typedef unsigned short bool;
+
 typedef signed long     int32;
 typedef unsigned long   uint32;
+#define MAX_UINT32 4,294,967,295L
```
See above, can't have commas in int literals

timburke commented on momo_modules/shared/portable/core/common_types.h:
```diff
@@ -10,10 +10,18 @@ typedef unsigned short bool;
 
 typedef signed char 	int8;
 typedef unsigned char	uint8;
```
We should just redefine these in terms of their corresponding types in stdint.h.  Those types are guaranteed to be the right side so we don't have to worry about int changing size on us if we switch chips.  See include/stdint.h in the directory where xc16 is installed.

amcgee replied:
Your sentence appears incomplete.  I believe both xc8 and xc16 have these sizes defined correctly.

timburke replied:
Sorry my angle brackets got de-html'ed.  I agree that it's fine.  It's just that we're spending time defining our own uint{8,16,32,64} when stdint.h is guaranteed by the ANSI standard to have uint{8,16,32,64}_t defined for us and correctly with the min and max values as well.

amcgee replied:
If I remember pullint in stdint was large, but I might be wrong.  That was
a long time ago.

On Sat, Apr 12, 2014 at 2:30 PM, timburke notifications@github.com wrote:

> In momo_modules/shared/portable/core/common_types.h:
> 
> > @@ -10,10 +10,18 @@ typedef unsigned short bool;
> > 
> >  typedef signed char    int8;
> >  typedef unsigned char  uint8;
> 
> Sorry my angle brackets got de-html'ed. I agree that it's fine. It's just
> that we're spending time defining our own uint{8,16,32,64} when stdint.h is
> guaranteed by the ANSI standard to have uint{8,16,32,64}_t defined for us
> and correctly with the min and max values as well.
> 
> ## 
> 
> Reply to this email directly or view it on GitHubhttps://github.com/WellDone/MoMo-Firmware/pull/66/files#r11562094
> .

## 

Austin McGee
austinwm@gmail.com
(510) 200-8486

timburke replied:
Could be.  I'm not opposed to keeping the code as is.  It works and the defines are the same as in stdint anways so its not a big deal.

timburke commented:
Looks good.  Only comments as noted above.  Let me know what you think and we can merge this.

amcgee commented:
I found a bug with flash_queue overrun, fixing now.

amcgee committed:
Fix overflow bug and add more functionality to momosensor (including a sensor event api feature test at )

amcgee committed:
Missed some files in the last commit

amcgee committed:
And some more files...

amcgee committed:
Finalize buffer overrun fix, implement an exhaustive test.

amcgee committed:
Expand the sensor event log back out to the full size (rather than just 2 subsections for testing).  Test will take forever.

amcgee commented:
Fixed a bunch of stuff and implemented an exhaustive test which can by run with `momosensor test`.  This tests writing values, overflowing the buffer (and subsequent subsections which contain values to be overwritten), reading past the end of the buffer, and more.  Note that running the test without reducing the size of the sensor event log (in memory_manager.c) will take forever, if you reduce the subsection count to 2 and run it it's reasonable and should hit all cases.

amcgee commented:
I haven't tested this with the sleep code, there could potentially be issues with response time.

amcgee committed:
Merge branch 'dev' into sensor-api

Conflicts:
	momo_modules/mainboard/src/momo/memory_manager.c
	tools/pymomo/commander/commands/rpc.py

amcgee commented:
Tested with the sleep changes, everything seemed to work fine and dandy.  Merged those changes here so this can be auto-merged now.
